# Lecture 16: Problem Set

##### Please note:

This problem, while given in Lecture 16, is also part of Lab 3, so don't
forget it when you submit the solutions to Lab 3.

In [1]:
from IPython.core.display import HTML
def css_styling():
    styles = open("../styles/tma4215.css", "r").read()
    return HTML(styles)

# Comment out next line and execute this cell to restore the default notebook style 
#css_styling()

In [19]:
import numpy as np
import pandas as pd

$$
\DeclareMathOperator{\Div}{div}
\DeclareMathOperator{\Grad}{grad}
\DeclareMathOperator{\Curl}{curl}
\DeclareMathOperator{\Rot}{rot}
\DeclareMathOperator{\ord}{ord}
\DeclareMathOperator{\Kern}{ker}
\DeclareMathOperator{\Image}{im}
\DeclareMathOperator{\spann}{span}
\DeclareMathOperator{\rank}{rank}
\DeclareMathOperator{\dist}{dist}
\DeclareMathOperator{\diam}{diam}
\DeclareMathOperator{\sig}{sig}
\DeclareMathOperator{\Id}{Id}
\newcommand{\RR}{\mathbb{R}}
\newcommand{\NN}{\mathbb{N}}
\newcommand{\VV}{\mathbb{V}}
\newcommand{\dGamma}{\,\mathrm{d} \Gamma}
\newcommand{\dGammah}{\,\mathrm{d} \Gamma_h}
\newcommand{\dx}{\,\mathrm{d}x}
\newcommand{\dy}{\,\mathrm{d}y}
\newcommand{\ds}{\,\mathrm{d}s}
\newcommand{\dt}{\,\mathrm{d}t}
\newcommand{\dS}{\,\mathrm{d}S}
\newcommand{\dV}{\,\mathrm{d}V}
\newcommand{\dX}{\,\mathrm{d}X}
\newcommand{\dY}{\,\mathrm{d}Y}
\newcommand{\dE}{\,\mathrm{d}E}
\newcommand{\dK}{\,\mathrm{d}K}
\newcommand{\dM}{\,\mathrm{d}M}
\newcommand{\cd}{\mathrm{cd}}
\newcommand{\onehalf}{\frac{1}{2}}
\newcommand{\bfP}{\boldsymbol P}
\newcommand{\bfx}{\boldsymbol x}
\newcommand{\bfy}{\boldsymbol y}
\newcommand{\bfa}{\boldsymbol a}
\newcommand{\bfu}{\boldsymbol u}
\newcommand{\bfv}{\boldsymbol v}
\newcommand{\bfe}{\boldsymbol e}
\newcommand{\bfb}{\boldsymbol b}
\newcommand{\bfc}{\boldsymbol c}
\newcommand{\bfq}{\boldsymbol q}
\newcommand{\bfy}{\boldsymbol y}
\newcommand{\bff}{\boldsymbol f}
\newcommand{\bfp}{\boldsymbol p}
\newcommand{\bft}{\boldsymbol t}
\newcommand{\bfj}{\boldsymbol j}
\newcommand{\bfB}{\boldsymbol B}
\newcommand{\bfV}{\boldsymbol V}
\newcommand{\bfE}{\boldsymbol E}
\newcommand{\bfB}{\boldsymbol B}
\newcommand{\bfzero}{\boldsymbol 0}
$$

## Exercise 1

For the interpolationspolynom $P(f|x_0, \ldots, x_n)$, prove the following recursion formlua:
\begin{align}
P(f|x_0, \ldots, x_n)(x) =
\dfrac{(x_0 - x)P(f|x_1, \ldots, x_n)(x) - (x_n - x)P(f|x_0, \ldots, x_{n-1})(x)}
{x_0 - x_n}
\end{align}


__Hint__: Show that the expression on the right hand side is a polynom $\tilde{P}$ of order $n$
which satfisfies the interpolation conditions $\tilde{P}(x_i) = f(x_i)$ for $i=0,\ldots n$. 
Why does that prove the recursion formula?

<font color='blue'>
    
   Your comments here: 

</font><br>
Denote the expression on the right hand side of the recursion formula above as $\tilde{P}(x_i)$. 

As $P(f|x_1,...,x_n)(x)$ and $P(f|x_0,...,x_{n-1})$ are interpolation polynomials of order $n-1$, $\tilde{P}(x)$ is of order $n$


\begin{aligned}
P(f|x_1,...,x_n)(x_i)&=P(f|x_0,...,x_{n-1})(x_i)\qquad i=1,...,n-1\\
\Rightarrow\tilde{P}(x_i)&=f(x_i)
\end{aligned}
For $i=0,n$ it is easily seen that $\tilde{P}(x_0)=f(x_0)$ and $\tilde{P}(x_n)=f(x_n)$. So $\tilde{P}(x_i)$ is a polynomial of order $n$ which interpolates $f$ at the $n+1$ nodes $x_0,...,x_n$. Theorem 7.1 in $\textbf{YEYEB}$ ensures the uniqueness of $P(f|x_0,...,x_n)(x)$.


## Exercise 2 (The Aiken-Neville Algorithm)

__a)__ Define the polynom $P_{ik}$ by
$$
P_{ik}(x) := P(f|x_{i-k}, \ldots,x_i)(x) \quad \text{for } i \geqslant k. 
$$
Then it follows from (1) that the value of $P_{nn} = P(f|x_0, \ldots, x_n)$ at some given point $x$
can be computed recursively as follows
\begin{alignat}{3}
P_{i0}(x) &= f_i \quad &&\text{for } i = 0,\ldots, n,
\\
P_{ik}(x) &= P_{i,k-1} + \dfrac{x - x_i}{x_i - x_{i-k}}(P_{i,k-1}(x) - P_{i-1,k-1}(x))
\quad && \text{for } i \geqslant k.
\end{alignat}

**Before you procede**, convince yourself that Equation (2) is just an application of the Equation (1) in Exercise 1 to $P(f|x_{i-k}, \ldots,x_i)(x)$.

This leads to the Aiken-Neville scheme (leaving out the function argument $x$):

\begin{equation}
\begin{matrix}
P_{00} & & & & & & 
\\
& \searrow &&&&&
\\
P_{10} & \rightarrow & P_{11} & & & &
\\
\vdots & & & \ddots & & &
\\
P_{n-1,0}& \rightarrow & \ldots & \rightarrow & P_{n-1,n-1} & \searrow &
\\
P_{n,0}& \rightarrow & \ldots & \rightarrow & P_{n-1,n-1} & \rightarrow & P_{nn}
\end{matrix}
\end{equation}

Based on this recursion, implement an ```aiken_neville``` function 
which takes in a function ```f```, a list of nodal points ```x_nodes``` 
and a value ```x``` and returns the value $P(f|x_0, \ldots, x_n)(x)$.

In [14]:
def aiken_neville(f, x_nodes, x):
    n = len(x_nodes)
    PX = np.zeros([n, n], dtype=float)
    for i in range(n):
        for k in range(i+1):
            if k==0:
                PX[i, k] = f(x_nodes[i])
            else:
                PX[i, k] = (PX[i, k-1] +
                            ((x - x_nodes[i])/(x_nodes[i] - x_nodes[i-k]))*(PX[i, k-1] - PX[i-1, k-1]))
    return PX

__b)__ Now use this function to compute an approximation of $\sin(62^{\circ})$
based on the data points 
$(50^{\circ},\sin(50^{\circ})),
(55^{\circ},\sin(55^{\circ})),
(60^{\circ},\sin(60^{\circ})),
\ldots,
(60^{\circ},\sin(70^{\circ}))$.

In [27]:
f = lambda x: np.sin(np.radians(x))
x_nodes = np.linspace(50,70,5)
x = 62
results = aiken_neville(f, x_nodes, x)
result_df = pd.DataFrame(results, columns=['sin(x_i)'] + ['' for i in range(len(x_nodes)-1)],
                        index=[str(int(deg)) for deg in x_nodes])
print(result_df)

    sin(x_i)                                        
50  0.766044  0.000000  0.000000  0.000000  0.000000
55  0.819152  0.893503  0.000000  0.000000  0.000000
60  0.866025  0.884775  0.883029  0.000000  0.000000
65  0.906308  0.882138  0.882929  0.882949  0.000000
70  0.939693  0.886277  0.882966  0.882946  0.882948


The ```aiken_neville``` function approximates $\sin(62^{\circ})\approx 0.882948$, this is very close to the true value

In [28]:
np.sin(np.radians(62))

0.8829475928589269